db/reqlog.c

```diff
@@ -2339,16 +2339,72 @@ static int hash_free_element(void *elem, void *unused) {
     return 0;
 }
 
+void acquire_clientstats_lock(int write)
+{
+    if (write) {
+        Pthread_rwlock_wrlock(&clientstats_lk);
+    } else {
+        Pthread_rwlock_rdlock(&clientstats_lk);
+    }
+}
+
+void release_clientstats_lock()
+{
+    Pthread_rwlock_unlock(&clientstats_lk);
+}
+
+static void init_clientstats(nodestats_t *entry, int task_len, char *host, int fd)
+{
+    entry->task = entry->mem;
+    entry->stack = entry->mem + task_len;
+    entry->host = host;
+    entry->ref = 1;
+    entry->rawtotals.api_history = init_api_history(); 
+    entry->rawtotals.svc_time = time_metric_new("svc_time");
+    entry->prevtotals.svc_time = entry->rawtotals.svc_time;
+    
+    Pthread_mutex_init(&entry->mtx, 0);
+    Pthread_mutex_init(&entry->rawtotals.lk, NULL);
+
+    if (fd < 0) {
+        bzero(&(entry->addr), sizeof(struct in_addr));
+    } else {
+        struct sockaddr_in peeraddr;
+        socklen_t len = sizeof(peeraddr);
+        bzero(&peeraddr, sizeof(peeraddr));
+        if (getpeername(fd, (struct sockaddr *)&peeraddr, &len) < 0) {
+            if (errno != ENOTCONN)
+                logmsg(LOGMSG_ERROR, "%s: getpeername failed fd %d: %d %s\n", __func__, fd, errno,
+                        strerror(errno));
+            bzero(&(entry->addr), sizeof(struct in_addr));
+        } else {
+            memcpy(&(entry->addr), &peeraddr.sin_addr,
+                    sizeof(struct in_addr));
+        }
+    }
+}
+
+static void update_clientstats_lru(nodestats_t *entry) {
+    if (entry->ref < 0) {
+        logmsg(LOGMSG_ERROR, "already a candidate for eviction.\n");
+        return;
+    }
+    
+    Pthread_mutex_lock(&clntlru_mtx);
+    listc_rfl(&clntlru, entry);
+    if (entry->ref) {
```

Comment:
I think you would always add it to the bottom of the lru list (we evict from the front). Say max is 2, client A inits (refcount initialized to 1), releases (refcount back to 0), and gets added to the top of the lru list; client B inits, releases, and gets added to the top of list too. Now client C comes in, instead of evicting A, which was used before B, we end up evicting B ...

thoughts?

Reply:
Does it matter whether A gets freed or B gets freed? They're both unused at that point. Unless we care about the sequential ordering of freed nodes, this is the simplest way to ensure that we always only release nodes with no references

Reply:
It would not be LRU anymore. This check `if (old_entry && old_entry->ref < 1)` of yours ensures that we don't evict nodestats that are still referenced.

Reply:
If it's traditional LRU then it'll have to iterate through the list of nodes until we find one we can evict since we only want to evict nodes with no refs remaining. For example:

1. client hits A
2. client hits B
3. client releases B
4. client hits C but A is the top node

Reply:
This IS an LRU.

Reply:
What are your thoughts about using a priority queue instead?

We're not benefiting from the current implementation and it doesn't seem like LRU actually fits our use case since we want to evict nodes with 0 references which is not guaranteed to be the oldest node.

Reply:
In all honesty I think we're over-optimizing this code :)

Reply:
After discussing with @mponomar , the cache should only be a lru cache of free-able nodes. Updated PR to reflect this.